### DISCUSSION TITLE
How to show number in specific format only?

### DISCUSSION THREAD

**Discussion by mianaliasjad** [vestrel00/contacts-android#256] [Q&A] [answered]
A RawContact has `List<Phone>`. Sometimes that list contains the same phone number but in different formats. For example, it can be "+923404140523" or "03404140523" or "0340 4140523"

They all are the same number. I want to show the number to the user in my Contacts app with "0340 4140523" this format only. 

How do I achieve that?

**Comment by vestrel00** [comment, marked as answer]
Hello @mianaliasjad! Thanks for being a user of this library 😁 I converted the issue you raised, #255, to this Q&A discussion because I believe that number formatting is more the responsibility of the application than this library.

> I want to show the number to the user in my Contacts app with "0340 4140523" this format only. How do I achieve that?

To answer this question effectively, we should look at how phone numbers are displayed to the user and stored in the Contacts provider database in the AOSP Contacts app, Google Contacts app, and Samsung Contacts app.

### What typical contacts applications do with phone numbers

In the following recordings, I took a look at how these Contacts apps behave when the user enters the same phone number in the same format and in different formats.

| AOSP Contacts                                                                                                          | Google Contacts                                                                                                          | Samsung Contacts                                                                                                          |
|------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------|
| ![aosp-contacts](https://user-images.githubusercontent.com/4332753/186955428-ea9d4d18-7f4f-468e-bbf4-4a904c4572a1.gif) | ![google-contacts](https://user-images.githubusercontent.com/4332753/186955412-7a3c3126-412b-40ab-9522-919e8ed7b388.gif) | ![samsung-contacts](https://user-images.githubusercontent.com/4332753/186955431-f672d09b-f0d9-4fe6-8f6b-d31e8b16ed3e.gif) |

As you can see, all contacts applications allow users to put the same number in different formats.I did not record this part but they will also allow you to enter the same exact number in the same exact format. Actually, if you watch closely at the end of the Samsung Contacts recording the last two phone numbers that were saved in a different format (space vs no space) are saved in the same format (no space).

I also checked the Contacts Provider data table and the duplicate phone numbers do exist. 

What we can conclude from this investigation is that it is up to the contacts applications (including users of this library) to perform formatting and duplicate checks if they (you) want. 

### What (you) can do with phone numbers

It is 100% up to you how you want to display phone numbers. This library will simply provide you easy read/write access to the Contacts Provider database.

I recommend you take a look at `PhoneNumberUtils`; https://developer.android.com/reference/android/telephony/PhoneNumberUtils

Here are some of the things you can do with it;

```kotlin
import android.telephony.PhoneNumberUtils

// "PAK" is the ISO-3166 code for Pakistan numbers (+92).
val countryIso = "PAK"
// In production, you should use the following to determine the ISO at runtime
// val telephonyManager = getSystemService(TELEPHONY_SERVICE) as TelephonyManager
// val countryIso = telephonyManager.networkCountryIso

Log.d("FORMATTED_NUMBER", PhoneNumberUtils.formatNumberToE164("+923404140523", countryIso).toString())
Log.d("FORMATTED_NUMBER", PhoneNumberUtils.formatNumber("3404140523").toString())
Log.d("FORMATTED_NUMBER", PhoneNumberUtils.formatNumber("340 4140523").toString())

if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
    Log.d("ARE_SAME_NUMBER", PhoneNumberUtils.areSamePhoneNumber("+923404140523", "03404140523", countryIso).toString())
    Log.d("ARE_SAME_NUMBER", PhoneNumberUtils.areSamePhoneNumber("0340 4140523", "03404140523", countryIso).toString())
} else {
    Log.d("ARE_SAME_NUMBER", PhoneNumberUtils.compare("+923404140523", "03404140523").toString())
    Log.d("ARE_SAME_NUMBER", PhoneNumberUtils.compare("0340 4140523", "03404140523").toString())
}
```

The above logs will print...

```
D/FORMATTED_NUMBER: +923404140523
D/FORMATTED_NUMBER: 340-414-0523
D/FORMATTED_NUMBER: 340 4140523

D/ARE_SAME_NUMBER: true
D/ARE_SAME_NUMBER: true
```

There are a lot of functions in `PhoneNumberUtils` that you can play around with.

Good luck! I hope this helps! If it answers your question, please mark this as the answer 😁

**Reply by mianaliasjad** [reply]
Thank you so much for detailed answer.

The duplicate records are coming against one contact.

So i can check if they are same with you provided method. Or i can convert all the numbers to e164 format and remove duplicates in that way.

Thanks again 🙂

**Reply by vestrel00** [reply]
You are welcome! Glad to have been able to help you 😁
